### PR TITLE
Edit snapshot

### LINKED ISSUE #4281: Add Edit snapshot capability to all snapshot views

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #4281

## Description

**TLDR;**
Introduces a common dialog and supporting endpoint for editing snapshots.

**Implementation**
* All snapshot table views (instance, application, device) now have the kebab menu item "Edit Device".
* When clicked, a form with `Name` and `Description` are pre-populated with the current values from the snapshot.
* Validation
   * Name has non empty validation only
   * "Confirm" button should be disabled when name is invalid
   * Description has no validation (can be blank)
* Upon clicking Confirm, the API is called with the new Name/Description
   * upon success the form is closed, a "success" toast is shown, and the table is updated (client side only)
   * upon failure, an "error" (warning) toast is shown and the dialog remains open to permit a retry without having to re-type

### Finer detail
* Adds `put` to the main route of the top-level snapshot API
  * expects `data` to contain `name` and/or `description` only at this time
* Implements `updateSnapshot` in controller `forge/db/controllers/Snapshot.js`
* Adds permission `snapshot:edit` that is used for both FE and BE routes
* Adds FE API route `updateSnapshot` to `frontend/src/api/snapshots.js`
* Adds shared dialog `frontend/src/components/dialogs/SnapshotEditDialog.vue`
* Updates kebab in the instances, devices and applications snapshots views (see demo GIF)

### Unit Tests added in `test/unit/forge/routes/api/snapshots_spec.js`
```
    Update snapshot
      instance
        ✔ Returns 404 for non-existent snapshot
        ✔ Non-member cannot update snapshot (72ms)
        ✔ Owner can update snapshot (52ms)
        ✔ Can update name only (51ms)
        ✔ Can update description only (53ms)
        ✔ Member cannot update snapshot (39ms)
      device
        ✔ Returns 404 for non-existent snapshot
        ✔ Non-member cannot update snapshot (44ms)
        ✔ Owner can update snapshot
        ✔ Can update name only
        ✔ Can update description only
        ✔ Member cannot update snapshot
```

### e2e Test added in `test/e2e/frontend/cypress/tests/applications/snapshots.spec.js`
```
FlowForge - Application - Snapshots
  ✔ provides functionality to edit a snapshot
```

### e2e Test added in `test/e2e/frontend/cypress/tests/instances/snapshots.spec.js`
```
FlowForge - Instance Snapshots
  ✔ provides functionality to edit a snapshot
```

### e2e Test added in `test/e2e/frontend/cypress/tests-ee/devices/snapshots.spec.js`
```
FlowForge - Devices - With Billing
  ✔ provides functionality to edit a snapshot
```

### Demo
chrome_qOr84XRdzp

### TODO
- [x] Tests (unit and e2e)

## Related Issue(s)

Owner #4281
Parent #3818

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by joepavitt** [comment]
How does this impact any snapshots deployed to a device? Is there a need to push updates at that point? I'm assuming we're just running from Snapshot ID, so no need?

**Comment by Steve-Mcl** [comment]
> How does this impact any snapshots deployed to a device? Is there a need to push updates at that point? I'm assuming we're just running from Snapshot ID, so no need?

Hmmm, I was 100% certain there was no need to update the device until you mentioned this 🤔 

_scurries off to double check_

**Comment by Steve-Mcl** [comment]
> > How does this impact any snapshots deployed to a device? Is there a need to push updates at that point? I'm assuming we're just running from Snapshot ID, so no need?
> 
> Hmmm, I was 100% certain there was no need to update the device until you mentioned this 🤔
> 
> _scurries off to double check_

@joepavitt thanks for flagging this.

So, the device does get `FF_SNAPSHOT_NAME` and as this PR stands, the device is not informed/not restarted meaning the env var is out of step.

This is however purely informational, and does not impact operation of the device.

The question becomes, do we _need_ to perform a device update just because of name change?  If we do, then it would be advisable to consider the UX (e.g. do we check ALL devices/instances associated with the snapshot, do we issue restart device automatically?)

It might be as well to toast info to the user (and update docs) to state something like:
_"Renaming a snapshot that is in use will not cause device to automatically restart however they will need an update at some point to synchronise the `FF_SNAPSHOT_NAME` environment variable."_

@knolleary any thoughts please?

**Comment by knolleary** [comment]
Agreed this *should not* trigger a device restart - if I'm doing housekeeping to mark a snapshot as 'good', I don't want it to redeploy and restart everything.

I would not use a toast notification either. It can be a simple one-line in the dialog (so the user knows this information *before* they make the change) to say something like "Changes to the snapshot will not be reflected on devices already running the snapshot".

**Comment by Steve-Mcl** [comment]
Note to reviewers: 

the TODO for audit logging [in lines 195-201](https://github.com/FlowFuse/flowfuse/pull/4282/files#diff-b0a947e68f7754337961ad87c95a14ba4565d8ffbdbbda9e5bb3a7713e8a0608R195-R201) of this PR are handled in the incoming branch PR #4287 [here](https://github.com/FlowFuse/flowfuse/pull/4287/files#diff-b0a947e68f7754337961ad87c95a14ba4565d8ffbdbbda9e5bb3a7713e8a0608R196-R209)

**Comment by Steve-Mcl** [comment]
> would not use a toast notification either. It can be a simple one-line in the dialog (so the user knows this information _before_ they make the change) to say something like "Changes to the snapshot will not be reflected on devices already running the snapshot".

Will do that. Thanks.

**Comment by Steve-Mcl** [comment]
Just hit a speed bump.

image

### Issue:

In short, because I renamed an auto snapshot, its name no longer matches the prescribed format and we can no longer infer this is a system auto snapshot.

We currently infer a snapshot was an auto snapshot due to lack of flags in db  (this has been mentioned in previous GH issue as being an MVP and will require a DB flag in the future).  It may be that future is now OR we can "handle it"

### Options (from simplest to hardest)
1. Leave it blank - size 0
2. Editor inherits the snapshot (current user info gets applied) - size 1
3. We update the database to have a field that records the originator or source of snapshot - size 2~3


Switching PR to draft for now until solution is determined/implemented.

**Comment by hardillb** [comment]
Gut feel is that if you edited, you own it...

**Comment by knolleary** [comment]
What other scenarios mean a snapshot has a null user property? Auto-generated snapshots is the only one I can think of - so why not use the lack of a user on the object to be the trigger to show the 'auto snapshot' label rather than parsing the name?

My only pause for thought is I cannot remember if our MVP for auto-snapshots included deleting old ones - and what logic that uses to determine what is an old auto-snapshot versus a snapshot a user wants to keep.

**Comment by Steve-Mcl** [comment]
> What other scenarios mean a snapshot has a null user property?

Stretching my memory here but 99.9% certain there should NOT be any operations other than auto snapshot that are permitted to leave `UserId` `null`.  My only pause is that I may have a distant memory of some snapshots already having no UserId (though that is very possibly due to local dev work)



> Auto-generated snapshots is the only one I can think of - so why not use the lack of a user on the object to be the trigger to show the 'auto snapshot' label rather than parsing the name?

That is an option (though it does move from "infer" to "assume")


> My only pause for thought is I cannot remember if our MVP for auto-snapshots included deleting old ones - and what logic that uses to determine what is an old auto-snapshot versus a snapshot a user wants to keep.

The premise was: if the name matches the specific format, there are 10 of them, delete the one with lowest `id` 

What will happen here, after an auto snapshot is renamed, there will now effectively be only 9 Auto snapshots. The next auto snapshot will not cause a deletion (but rather make up the 10th) - so this is a non issue.




> Gut feel is that if you edited, you own it...

I am slightly leaning to this also but there is merit in what Nick said about assuming `null` `UserId` as an Auto Snapshot - as keeps the status quo!


Since modifying the DB to identify snapshot source (discussed [here ](https://github.com/FlowFuse/flowfuse/issues/3358#issuecomment-1935790720) and [here](https://github.com/FlowFuse/flowfuse/issues/3355#issuecomment-1948246802)) are not absolutely necessary at this point (only auto snapshots apply `null` UserId) we can move move forwards with the smaller iteration at this time.  Therefore, could I get consensus on the 2 offerings please: 

1. Assume any snapshot NOT matching the name pattern and has `null` `UserId` is an Auto Snapshot
2. User who changes, now owns it

**Comment by knolleary** [comment]
Please go with my suggestion. The user field is who created it, not who owns it. If the field is blank, assume its an auto-snapshot - ie the system generated it. That is a safe assumption to make - *unless* the user who created has since been deleted. But that is an edge case for another day.

**Comment by Steve-Mcl** [comment]
> I would not use a toast notification either. It can be a simple one-line in the dialog (so the user knows this information _before_ they make the change) to say something like "Changes to the snapshot will not be reflected on devices already running the snapshot".

Addressed in [937807d](https://github.com/FlowFuse/flowfuse/pull/4282/commits/937807d60fbe083e05cfbbf9dec29343c7238d59)

image

**Comment by Steve-Mcl** [comment]
> If the field is blank, assume its an auto-snapshot - ie the system generated it.

Addressed in [d46a226](https://github.com/FlowFuse/flowfuse/pull/4282/commits/d46a2267e10878b9a8b70307fb2988760de40978)

image

**Review by joepavitt — APPROVED** [review]
FE all good. Server-side looks good too, but will wait for Nick's eyes before merging.

**Comment by knolleary** [comment]
I'm not going to be able to review this properly until much later today.

**Review by knolleary — APPROVED** [review]

**Comment by knolleary** [comment]
Merging once the tests go green
